### PR TITLE
[string] Clean up StringObject; Clarify its tagged BridgeObjects.

### PR DESCRIPTION
**Pull request by milseman**
Bifurcate StringObjects raw bit-pattern initializers to expose whether
the caller is passing a value rather than a reference. This way,
StringObject can call the approprite BridgeObject helper and
participate in upcoming peephole optimizations like retains of
known-values. This is all meant to be a NOP.

Additionally, do some cleanup while we're at it.

### CODE REVIEW

File: stdlib/public/core/StringObject.swift [hunk 1]
```diff
@@ -96,13 +100,45 @@ extension _StringObject {
   @inline(__always)
   // TODO: private
   internal
-  init(rawBits: _RawBitPattern) {
+  init(taggedRawBits: _RawBitPattern) {
+#if arch(i386) || arch(arm)
+    self.init(Builtin.reinterpretCast(taggedRawBits), taggedRawBits &>> 32)
```
Comment (lorentey) [inline stdlib/public/core/StringObject.swift]: Hm, interesting. I would've expected both of these parameter expressions to blow up. (Due to UInt64 and _Variant not having the same size in reinterpretCast, and the UInt64/UInt type mismatch with &>>. Clearly I’m missing something obvious, but what?

File: stdlib/public/core/StringObject.swift [hunk 2]
```diff
@@ -340,7 +376,7 @@ extension _StringObject {
   @inline(__always)
   internal
   init() {
-    self.init(rawBits: UInt64(_StringObject._emptyStringBitPattern))
+    self.init(taggedRawBits: _StringObject._emptyStringBitPattern)
```
Comment (lorentey) [inline stdlib/public/core/StringObject.swift]: This won't compile on 32-bit platforms.

Comment (milseman) [reply stdlib/public/core/StringObject.swift]: Right!

Comment (milseman) [reply stdlib/public/core/StringObject.swift]: And by "right" I mean wrong! At least for this code which is inside a `#if`. The inits certainly won't compile, though.

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test compiler performance

**Comment by milseman** [comment]
(to help see if there was a consistent pattern of regression vs noises)

@swift-ci please smoke test

**Comment by milseman** [comment]
(to help see if there is a consistent regression or just noise)

@swift-ci please smoke benchmark

File: stdlib/public/core/StringObject.swift (hunk 2 above)
Comment (lorentey) [reply stdlib/public/core/StringObject.swift]: 🤦‍♂️ Wow, I didn't expand far enough 😊

**Review by lorentey — APPROVED** [review]
Looking good

**Comment by milseman** [comment]
@graydon / @DougGregor / @xedin / @rudkx , do you know what the Sema.NumConstraintScopes metric is tracking and how this could of regressed?

**Comment by DougGregor** [comment]
`Sema.NumConstraintScopes` is a measure of how many solution states the type checker solves. It could regress when, say, new overloads are added to the standard library.

**Comment by milseman** [comment]
Are two functions with the same base name but different labels (i.e. different full names) considered overloads? In this PR, I added some internal methods that have different full names.

**Pull request merged**